README.md
# proxysnps

`proxysnps` is an R package that implements functions to get proxy SNPs
in [linkage disequilibrium][LD] (LD) with a [SNP] in the [1000 Genomes
Project][1000genomes].

```r
library(proxysnps)
d <- get_proxies(query = "rs42")
plot(d$POS, d$R.squared, main="rs42", xlab="Position", ylab=bquote("R"^2))
```

![rs42](https://github.com/slowkow/proxysnps/blob/master/rs42.png)

See the [vignette] for more usage examples.

## Installation

```r
install.packages("devtools")
devtools::install_github("slowkow/proxysnps")
```

## Data

This package provides easy access to 1000 Genomes Project VCF files that have
been filtered by [Brian Browning][browning], available
[here][beagle_1000genomes].

[browning]: https://faculty.washington.edu/browning/
[beagle_1000genomes]: http://bochet.gcc.biostat.washington.edu/beagle/1000_Genomes_phase3_v5a/

## Contributing

Please [submit an issue][issues] to report bugs or ask questions.

Please contribute bug fixes or new features with a [pull request][pull] to this
repository.

[issues]: https://github.com/slowkow/proxysnps/issues
[pull]: https://help.github.com/articles/using-pull-requests/

## Related work

[HaploReg][haploreg]

> HaploReg is a tool for exploring annotations of the noncoding genome at
> variants on haplotype blocks, such as candidate regulatory SNPs at
> disease-associated loci. Using [LD] information from the [1000 Genomes
> Project][1000genomes], linked SNPs and small indels can be visualized along
> with chromatin state and protein binding annotation from the [Roadmap
> Epigenomics][roadmap] and [ENCODE] projects, sequence conservation across
> mammals, the effect of SNPs on regulatory motifs, and the effect of SNPs on
> expression from [eQTL] studies.

[LDheatmap]

> Produces a graphical display, as a heat map, of measures of pairwise linkage
> disequilibria between SNPs. Users may optionally include the physical
> locations or genetic map distances of each SNP on the plot.

Also see the [Statistical Genetics CRAN Task View][statgen] for additional
R packages.

[LocusZoom][locuszoom]

> LocusZoom is a tool to plot regional association results from genome-wide
> association scans or candidate gene studies.

[SNAP][snap] (defunct)

> SNAP is a computer program and web-based service for the rapid retrieval of
> linkage disequilibrium proxy SNP results given input of one or more query
> SNPs and based on empirical observations from the [International HapMap
> Project][hapmap] and the [1000 Genomes Project][1000genomes].

[Tagger][tagger]

> Tagger is a tool for the selection and evaluation of tag SNPs from genotype
> data such as that from the [International HapMap Project][hapmap]. It
> combines the simplicity of pairwise tagging methods with the efficiency
> benefits of multimarker haplotype approaches.

[LD]: https://en.wikipedia.org/wiki/Linkage_disequilibrium
[SNP]: https://en.wikipedia.org/wiki/Single-nucleotide_polymorphism
[eQTL]: https://en.wikipedia.org/wiki/Expression_quantitative_trait_loci

[vignette]: https://github.com/slowkow/proxysnps/blob/master/vignettes/proxysnps.md

[hapmap]: http://www.hapmap.org/
[1000genomes]: http://www.1000genomes.org/
[GRCh37]: http://www.1000genomes.org/faq/which-reference-assembly-do-you-use
[roadmap]: http://www.roadmapepigenomics.org/
[ENCODE]: https://www.encodeproject.org/

[haploreg]: http://www.broadinstitute.org/mammals/haploreg/
[LDheatmap]: https://cran.r-project.org/web/packages/LDheatmap/index.html
[statgen]: https://cran.r-project.org/web/views/Genetics.html
[locuszoom]: http://locuszoom.sph.umich.edu/locuszoom/
[snap]: http://www.broadinstitute.org/mpg/snap/
[tagger]: https://www.broadinstitute.org/mpg/tagger/


CHANGELOG.md
# Change Log

# 0.0.2 - 2015-12-15

## Added

- Send queries to myvariant.info to get chromosome and position.
- Simplified example in `README.md`.

## Fixed

- `R/sysdata.rda` wasn't working so I switched to `data/ind.rda`

## Changed

- Change default `pop` to `NA` instead of `EUR`.

# 0.0.1 - 2015-12-15

- Initial release.


R/get_proxies.R
#' Get proxy SNPs for a SNP at a given genomic position.
#' 
#' Returns a dataframe with proxy SNPs.
#' 
#' Currently, this is hard-coded to access 1000 Genomes phase3 data hosted by
#' Brian Browning (author of BEAGLE):
#' 
#' \url{http://bochet.gcc.biostat.washington.edu/beagle/1000_Genomes_phase3_v5a/}
#' 
#' This implementation discards multi-allelic markers that have a "," in the
#' ALT column.
#'
#' The \code{pop} can be any of: ACB, ASW, BEB, CDX, CEU, CHB, CHS, CLM, ESN,
#' FIN, GBR, GIH, GWD, IBS, ITU, JPT, KHV, LWK, MSL, MXL, PEL, PJL, PUR, STU,
#' TSI, YRI. It can also be any super-population: AFR, AMR, EAS, EUR, SAS.
#' 
#' Find more details here:
#' \url{http://www.1000genomes.org/faq/which-populations-are-part-your-study}
#' 
#' @param chrom a chromosome name (1-22,X) without "chr"
#' @param pos a positive integer indicating the position of a SNP
#' @param window_size a positive integer indicating the size of the window
#' @param pop the name of a 1000 Genomes population (AMR,AFR,ASN,EUR,...). Set
#'        this to NA to use all populations.
#' @return A dataframe with the following columns:
#'   \describe{
#'     \item{CHROM}{Chromosome name, e.g. "1"}
#'     \item{POS}{Position, e.g. 583090}
#'     \item{ID}{Identifier, e.g. "rs11063140"}
#'     \item{REF}{Reference allele, e.g. "A"}
#'     \item{ALT}{Alternative allele, e.g. "G"}
#'     \item{MAF}{Minor allele frequency, e.g. 0.1}
#'     \item{R.squared}{Squared Pearson correlation coefficient, e.g. 1.0}
#'     \item{D.prime}{D prime value, e.g. 1.0}
#'     \item{CHOSEN}{Binary indicator set to TRUE for the SNP of interest}
#'   }
#' 
#' @examples
#' d <- get_proxies(chrom = "12", pos = 583090, window_size = 1e5, pop = "AFR")
#' head(d)
#'  
#' @export
get_proxies <- function(
  chrom = NA, pos = NA, query = NA, window_size = 1e5, pop = NA) {
  
#   chrom = "12"
#   pos = 583090
#   window_size = 1e5
#   pop = "AFR"
  
  # Try to get the position from myvariant.info if there's a query.
  if (is.na(chrom) && is.na(pos) && !is.na(query)) {
    v <- myvariant::queryVariant(query)
    if (v$total > 0) {
      chrom <- v$hits$dbsnp$chrom[1]
      pos <- v$hits$dbsnp$hg19$start[1]
    } else {
      stop(sprintf("[myvariant.info] No hits for query '%s'", query))
    }
  }
  
  # Get VCF data for this genomic region.
  vcf <- get_vcf(
    chrom = chrom, 
    start = floor(pos - window_size / 2),
    end =  floor(pos + window_size / 2),
    pop = pop)
  
  # Separate the metadata from the genotypes.
  meta <- vcf$meta
  geno <- vcf$geno
  
  # Drop unused columns.
  meta$QUAL <- NULL
  meta$FILTER <- NULL
  meta$INFO <- NULL
  
  # Compute the minor allele frequencies.
  maf <- sapply(rowSums(geno) / ncol(geno), function(x) min(x, 1 - x))
  
  # Select the SNP of interest.
  chosen <- meta$CHROM == chrom & meta$POS == pos
  
  # If the position does not match any SNPs, choose the middle one.
  if (!length(which(chosen))) {
    chosen <- rep(FALSE, nrow(meta))
    chosen[floor(nrow(meta) / 2)] <- TRUE
  }
  
  # Compute R.squared and D.prime with binary (0/1) markers.
  ld <- apply(geno, 1, function(y) {
    compute_ld(geno[which(chosen),], y)
  })
  
  # Create a dataframe.
  retval <- cbind(
    meta,
    MAF = maf,
    R.squared = round(sapply(ld, "[[", "R.squared"), 6),
    D.prime = round(sapply(ld, "[[", "D.prime"), 6),
    CHOSEN = chosen
  )
  
  return(retval[order(retval$R.squared, decreasing = TRUE),])
}

#' Compute two commonly used linkage disequilibrium statistics.
#' 
#' Compute R.squared and D.prime for two binary numeric vectors.
#' 
#' Find more details here:
#' \url{https://en.wikipedia.org/wiki/Linkage_disequilibrium}
#' 
#' @param x a numeric vector of ones and zeros
#' @param y a numeric vector of ones and zeros
#' @return A list with two items:
#' \describe{
#'  \item{R.squared}{Squared Pearson correlation coefficient.}
#'  \item{D.prime}{Coefficient of linkage disequilibrium D divided by the
#'   theoretical maximum.}
#' }
#' 
#' @examples
#' compute_ld(c(0,0,0,1,1,1), c(1,1,1,1,0,0))
#' 
#' @export
compute_ld <- function(x, y) {
  stopifnot(all(names(table(x)) %in% c("0", "1")))
  stopifnot(all(names(table(y)) %in% c("0", "1")))
  stopifnot(length(x) == length(y))
  
  # Returned list.
  retval <- list("D.prime" = 0, "R.squared" = 0)
  
  # Allele frequencies.
  px <- max(sum(x) / length(x), .Machine$double.eps)
  py <- max(sum(y) / length(x), .Machine$double.eps)
  pxy <- sum(x & y) / length(x)
  
  # D prime.
  d <- pxy - px * py
  dmax <- 0
  if (d < 0) {
    dmax <- min(px * py, (1 - px) * (1 - py))
  } else {
    dmax <- min(px * (1 - py), (1 - px) * py)
  }
  if (dmax != 0) {
    retval[["D.prime"]] <- d / dmax
  }
  
  # Squared Pearson correlation coefficient.
  retval[["R.squared"]] <- (d / sqrt(px * (1 - px) * py * (1 - py))) ^ 2
  
  return(retval)
}


data-raw/make_rdata.R
ind <- read.delim(
  file = "data-raw/20130606_g1k.ped.gz",
  stringsAsFactors = FALSE
)

# Delete unused columns.
for (col in colnames(ind)) {
  if (length(table(ind[[col]])) == 1) {
    ind[[col]] <- NULL
  }
}

rownames(ind) <- ind$Individual.ID

ind$Gender <- ifelse(ind$Gender == 2, "female", "male")

ind$SuperPopulation[ind$Population == "CHB"] <- "EAS"
ind$SuperPopulation[ind$Population == "JPT"] <- "EAS"
ind$SuperPopulation[ind$Population == "CHS"] <- "EAS"
ind$SuperPopulation[ind$Population == "CDX"] <- "EAS"
ind$SuperPopulation[ind$Population == "CHB"] <- "EAS"
ind$SuperPopulation[ind$Population == "KHV"] <- "EAS"

ind$SuperPopulation[ind$Population == "CEU"] <- "EUR"
ind$SuperPopulation[ind$Population == "TSI"] <- "EUR"
ind$SuperPopulation[ind$Population == "FIN"] <- "EUR"
ind$SuperPopulation[ind$Population == "GBR"] <- "EUR"
ind$SuperPopulation[ind$Population == "IBS"] <- "EUR"

ind$SuperPopulation[ind$Population == "YRI"] <- "AFR"
ind$SuperPopulation[ind$Population == "LWK"] <- "AFR"
ind$SuperPopulation[ind$Population == "GWD"] <- "AFR"
ind$SuperPopulation[ind$Population == "MSL"] <- "AFR"
ind$SuperPopulation[ind$Population == "ESN"] <- "AFR"
ind$SuperPopulation[ind$Population == "ASW"] <- "AFR"
ind$SuperPopulation[ind$Population == "ACB"] <- "AFR"

ind$SuperPopulation[ind$Population == "MXL"] <- "AMR"
ind$SuperPopulation[ind$Population == "PUR"] <- "AMR"
ind$SuperPopulation[ind$Population == "CLM"] <- "AMR"
ind$SuperPopulation[ind$Population == "PEL"] <- "AMR"

ind$SuperPopulation[ind$Population == "GIH"] <- "SAS"
ind$SuperPopulation[ind$Population == "PJL"] <- "SAS"
ind$SuperPopulation[ind$Population == "BEB"] <- "SAS"
ind$SuperPopulation[ind$Population == "STU"] <- "SAS"
ind$SuperPopulation[ind$Population == "ITU"] <- "SAS"

save(list = c("ind"), file = "data/ind.rda")


R/get_vcf.R
#' Get data for a genomic region from a remote VCF file.
#' 
#' Returns a list with three dataframes for individuals, SNPs, and genotypes.
#' 
#' Currently, this is hard-coded to access 1000 Genomes phase3 data hosted by
#' Brian Browning (author of BEAGLE):
#' 
#' \url{http://bochet.gcc.biostat.washington.edu/beagle/1000_Genomes_phase3_v5a/}
#' 
#' This implementation discards multi-allelic markers that have a "," in the
#' ALT column.
#' 
#' The \code{pop} can be any of: ACB, ASW, BEB, CDX, CEU, CHB, CHS, CLM, ESN,
#' FIN, GBR, GIH, GWD, IBS, ITU, JPT, KHV, LWK, MSL, MXL, PEL, PJL, PUR, STU,
#' TSI, YRI. It can also be any super-population: AFR, AMR, EAS, EUR, SAS.
#'
#' Find more details here:
#' \url{http://www.1000genomes.org/faq/which-populations-are-part-your-study}
#' 
#' @param chrom a chromosome name (1-22,X) without "chr"
#' @param start a positive integer indicating the start of a genomic region
#' @param end a positive integer indicating the end of a genomic region
#' @param pop the name of a 1000 Genomes population (AMR,AFR,ASN,EUR,...)
#' @return A list with three dataframes:
#' \describe{
#'   \item{ind}{A dataframe with information about individuals: Family.ID,
#'    Individual.ID, Paternal.ID, Maternal.ID, Gender, Population,
#'    Relationship, Siblings, Second.Order, Third.Order, Other.Comments,
#'    SuperPopulation}
#'   \item{meta}{First 8 columns of the VCF file: CHROM, POS, ID, REF, ALT,
#'    QUAL, FILTER, INFO}
#'   \item{geno}{Columns 10 onward of the VCF file. All genotypes are converted
#'    to 0s and 1s representing REF and ALT alleles. This dataframe has two
#'    columns for each individual.}
#' }
#' 
#' @examples
#' vcf <- get_vcf(chrom = "12", start = 533090, end = 623090, pop = "AFR")
#' names(vcf)
#'  
#' @export
get_vcf <- function(chrom, start, end, pop = NA) {
  
  # Hard-coded superpopulations for each individual.
  superpops <- rep(
    c("EUR","EAS","AMR","EAS","AMR","EAS","AMR","EAS","AMR","EUR","AMR",
      "EUR","AMR","SAS","EAS","EUR","EAS","AFR","AMR","AFR","AMR","AFR",
      "AMR","AFR","AMR","AFR","EAS","AFR","EAS","AMR","AFR","AMR","AFR",
      "EAS","AFR","AMR","EAS","EUR","EAS","AMR","AFR","AMR","AFR","AMR",
      "AFR","AMR","AFR","AMR","EAS","AFR","AMR","AFR","SAS","AFR","EAS",
      "AFR","SAS","AFR","SAS","AFR","SAS","AFR","SAS","AFR","SAS","AFR",
      "SAS","AFR","SAS","AFR","SAS","AFR","SAS","EUR","AFR","EAS","AFR",
      "EAS","AFR","EAS","AFR","AMR","AFR","AMR","AFR","EUR","SAS"),
    c(185,42,3,35,4,28,9,15,111,1,68,24,9,4,6,73,56,8,2,5,24,2,13,5,7,5,
      17,4,21,2,1,3,3,17,2,3,20,14,1,2,2,11,4,8,3,1,13,2,35,2,1,22,4,7,
      4,27,5,16,10,6,13,6,9,9,12,58,10,84,6,74,3,25,310,99,19,103,29,70,
      18,34,142,19,8,45,52,107,103)
  )
  
  # Hard-coded populations for each individual.
  pops <- rep(
    c("GBR","FIN","GBR","FIN","CHS","PUR","CHS","PUR","CHS","PUR","CDX","PUR",
      "CLM","PUR","CLM","PUR","GBR","CLM","PUR","CLM","IBS","CLM","PEL","PJL",
      "KHV","IBS","GBR","CDX","KHV","ACB","PEL","ACB","PEL","ACB","PEL","ACB",
      "PEL","ACB","KHV","ACB","KHV","PEL","ACB","PEL","ACB","KHV","ACB","PEL",
      "CDX","GBR","IBS","CDX","PEL","ACB","PEL","ACB","PEL","ACB","PEL","ACB",
      "PEL","CDX","ACB","PEL","ACB","GWD","ACB","PJL","ACB","KHV","ACB","GWD",
      "ACB","GWD","PJL","GWD","PJL","GWD","PJL","GWD","PJL","GWD","PJL","GWD",
      "ESN","GWD","BEB","PJL","GWD","MSL","ESN","MSL","PJL","GWD","ESN","MSL",
      "PJL","ESN","GWD","MSL","BEB","PJL","STU","PJL","STU","PJL","STU","ITU",
      "STU","ITU","STU","PJL","ITU","BEB","STU","ITU","STU","BEB","STU","ITU",
      "STU","ITU","STU","ITU","STU","ITU","STU","ITU","STU","ITU","BEB","ITU",
      "STU","ITU","STU","ITU","CEU","YRI","CHB","YRI","JPT","LWK","JPT","YRI",
      "LWK","ASW","MXL","ASW","MXL","ASW","TSI","GIH"),
    c(55,17,31,82,42,3,35,4,28,9,15,41,18,26,16,10,1,27,11,30,24,3,6,4,6,70,
      3,22,34,8,2,5,24,2,13,5,7,5,17,4,21,2,1,3,3,17,2,3,20,1,13,1,2,2,11,4,
      8,3,1,13,2,35,2,1,9,4,9,4,7,4,10,7,2,8,5,16,10,6,13,6,9,9,12,37,19,2,4,
      6,10,27,43,4,6,6,29,39,3,8,2,15,13,8,5,7,19,6,1,11,5,1,12,3,20,20,13,15,
      13,21,9,12,8,2,2,10,7,8,1,7,4,1,28,5,1,7,2,3,99,19,103,29,70,18,34,60,
      81,1,19,8,45,52,107,103)
  )
  
  if (!is.na(pop) && !pop %in% pops && !pop %in% superpops) {
    stop(
      "Invalid pop=", pop,
      "\nMust be a pop: ", paste(unique(pops), collapse = " "),
      "\nOr a superpop: ", paste(unique(superpops), collapse = " ")
    )
  }
  
  if (!is.numeric(start) || start < 1) {
    stop("Invalid start=", start, "\nMust be a positive integer.")
  }
  
  if (!is.numeric(end) || end < start) {
    stop("Invalid end=", end, "\nMust be an integer >= start.")
  }
  
  if (!chrom %in% c(1:22, "X")) {
    stop(
      "Invalid chrom=", chrom,
      "\nMust be one of: ", paste(c(1:22, "X"), collapse = " ")
    )
  }
  
  # These are variants filtered by Brian Browning, the developer of BEAGLE.
  data_url = paste(sep = "",
    "http://tabix.iobio.io/?cmd=-h%20%27",
    "http://bochet.gcc.biostat.washington.edu/beagle/1000_Genomes_phase3_v5a/",
    "individual_chromosomes/chr", chrom, ".1kg.phase3.v5a.vcf.gz",
    "%27%20", chrom, ":", start, "-", end
  )
  
  # Download the data from the server.
  txt <- RCurl::getURL(data_url)
  
  # Extract the sample identifiers from the VCF header.
  sample_ids <- strsplit(
    # FIXME: Should detect number of comment lines.
    # Get the 5th line and split it.
    strsplit(txt, "\n", fixed = TRUE)[[1]][5],
    "\t", fixed = TRUE
  )[[1]]
  
  # Discard the "CHROM,POS,...,INFO,V9" columns.
  sample_ids <- sample_ids[10:length(sample_ids)]
  
  # Read the body of the data into a dataframe.
  vcf <- read.delim(
    text = txt, header = FALSE, comment.char = "#", stringsAsFactors = FALSE)
  
  # Assign the standard column names and sample identifiers.
  colnames(vcf) <- c(
    "CHROM", "POS", "ID", "REF", "ALT", "QUAL", "FILTER", "INFO", "V9",
    sample_ids
  )
  
  # Discard multi-allelic markers.
  vcf <- vcf[grep(",", vcf$ALT, invert = TRUE), ]
  
  # Select the genotype columns that belong to a particular population.
  if (pop %in% pops) {
    vcf <- vcf[,c(rep(TRUE, 9), pops == pop)]
  } else if (pop %in% superpops) {
    vcf <- vcf[,c(rep(TRUE, 9), superpops == pop)]
  }
  
  retval <- list()
  
  #load("data/sysdata.rda")
  data("ind", package = "proxysnps", envir = environment())
  # retval$ind <- proxysnps::ind[colnames(vcf)[10:ncol(vcf)],]
  retval$ind <- ind[colnames(vcf)[10:ncol(vcf)],]
  
  # Separate the metadata from the genotypes.
  retval$meta <- vcf[,1:8]
  retval$geno <- vcf[,10:ncol(vcf)]
  
  # Convert the genotypes to a numeric matrix.
  retval$geno <- t(apply(retval$geno, 1, function(row) {
    as.numeric(do.call(cbind, strsplit(row, "|", fixed = TRUE)))
  }))
  
  rownames(retval$geno) <- retval$meta$ID
  colnames(retval$geno) <- rep(retval$ind$Individual.ID, each = 2)
  
  return(retval)  
}